<a href="https://colab.research.google.com/github/Mahnoor-Kalsoom/flyrank-ml-internship/blob/main/work/notebooks/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Mahnoor-Kalsoom/flyrank-ml-internship/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

Lane: Refresh / Content Opportunity Scoring

I chose the Refresh / Content Opportunity Scoring lane because I want to investigate which pages should be reviewed first for possible refresh, improvement, protection, or monitoring. This is useful because a content team usually has limited time and cannot manually review every page.

The starter dataset already provides observable signals such as impressions, sessions, content age, freshness, average position, CTR, and trend direction. I can use these signals to build and compare a transparent baseline with a simple ML ranking model.

I am treating this as a decision-support problem, not as a system that guarantees which page will improve after a refresh.

In [1]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

Research question

Which pages should be reviewed first for a potential content refresh, based on observable search and content signals?

Decision

The decision is which pages should receive limited content-review attention first.

Action

For high-priority pages, a reviewer could inspect the page and decide whether to:

refresh or update the content;
expand thin content;
investigate CTR issues;
protect a page that is performing well;
monitor it instead of making a change.
Cost of a wrong recommendation

A false positive could cause a team to spend time reviewing or changing a page that did not need attention.

A false negative could be more costly because a genuinely declining or valuable page might be missed.

Therefore, I want the ranking to prioritize useful candidates while keeping the reasoning understandable to a human reviewer.


In [2]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*



In [7]:
!git clone https://github.com/Mahnoor-Kalsoom/flyrank-ml-internship.git

Cloning into 'flyrank-ml-internship'...
remote: Enumerating objects: 130, done.
remote: Counting objects: 100% (130/130), done.
remote: Compressing objects: 100% (80/80), done.
remote: Total 130 (delta 43), reused 107 (delta 34), pack-reused 0 (from 0)
Receiving objects: 100% (130/130), 1.87 MiB | 5.13 MiB/s, done.
Resolving deltas: 100% (43/43), done.


In [8]:
%cd /content/flyrank-ml-internship

/content/flyrank-ml-internship


In [9]:
!ls

AGENTS.md  DATA_USE.md	LICENSE    README.md	     SETUP.md	 work
CLAUDE.md  docs		notebooks  requirements.txt  skills
data	   GUIDE.md	outputs    scripts	     submission


In [10]:
!ls data/raw

content_refresh_anonymized.csv


In [12]:
import pandas as pd

df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

df["is_declining_label"] = (
    df["trend_direction"].str.lower().eq("down").astype(int)
)

print("Number of pages:", len(df))
print("Declining-label rate:", round(df["is_declining_label"].mean(), 3))
print("Median impressions:", df["impressions_90d"].median())
print("Median content age:", df["content_age_days"].median())

Number of pages: 30000
Declining-label rate: 0.542
Median impressions: 731.0
Median content age: 236.0


## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

The starter dataset contains 30,000 pages and approximately 54.2% of pages have the starter is_declining_label.

The starter results also show that the simple baseline achieved Precision@50 of 0.240, while the random forest achieved Precision@50 of 0.740 under the starter's client-holdout validation.

This suggests that there may be useful signal beyond a simple hand-written refresh rule. However, these results are from the starter dataset and its proxy label, so I will not treat them as proof that the model can predict whether a real content refresh will succeed.


In [3]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.